samcli/lib/utils/git_repo.py

```diff
@@ -0,0 +1,180 @@
+""" Manage Git repo """
+
+import logging
+import os
+import platform
+import shutil
+import subprocess
+from pathlib import Path
+from typing import Optional
+
+from samcli.lib.utils import osutils
+from samcli.lib.utils.osutils import rmtree_callback
+
+LOG = logging.getLogger(__name__)
+
+
+class CloneRepoException(Exception):
+    """
+    Exception class when clone repo fails.
+    """
+
+
+class CloneRepoUnstableStateException(CloneRepoException):
+    """
+    Exception class when clone repo enters an unstable state.
+    """
+
+
+class GitRepo:
+    """
+    Class for managing a Git repo, currently it has a clone functionality only
+
+    Attributes
+    ----------
+    _url: str
+        The URL of this Git repository, example "https://github.com/aws/aws-sam-cli"
+    _name: str
+        The name of this Git repository, example "aws-sam-cli"
+    _local_path: Path
+        The path of the local clone of this Git repository
+    _clone_attempted: bool
+        whether an attempt to clone this Git repository took place or not
+
+    Methods
+    -------
+    clone(self, clone_dir: Path, clone_name=None, replace_existing=False) -> Path:
+        creates a local clone of this Git repository
+    """
```

Comment:
This class is pretty hard to understand, can we add more details in docstring about what it does and why we need `_local_path` and `_clone_attempted`. I think I sort of get the idea of "avoid cloning multiple times" but it is not clear about what roles these two attributes are playing.